### PR TITLE
Guidelines for aria-level attribute

### PR DESCRIPTION
**Pull request by spectranaut**
Initial draft of content for issue #259 to provide a guidance section on aria-level.

Might still need some discussion related to this open issue on ARIA: w3c/aria#915

***

### :boom: Error: 500 Internal Server Error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Dec 2, 2019, 9:27 AM UTC)_.

<details>
<summary>More</summary>

PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [HTML Diff Service](http://services.w3.org/htmldiff) - The HTML Diff Service is used to create HTML diffs of the spec changes suggested in a pull request.

:link: [Related URL](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fpr-preview.s3.amazonaws.com%2Fw3c%2Faria-practices%2Fpull%2F1109%2F0d79d21.html&doc2=https%3A%2F%2Fpr-preview.s3.amazonaws.com%2Fw3c%2Faria-practices%2Fpull%2F1109.html)

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>500 Internal Server Error</title>
</head><body>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error or
misconfiguration and was unable to complete
your request.</p>
<p>Please contact the server administrator at 
 sysreq@w3.org to inform them of the time this error occurred,
 and the actions you performed just before this error.</p>
<p>More information about this error may be available
in the server error log.</p>
</body></html>

```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20w3c/aria-practices%231109.)._
</details>

### CODE REVIEW

**Comment by zcorpan** [comment]
I don't think this section needs to be blocked on https://github.com/w3c/aria/issues/915 - we can update APG if necessary when ARIA changes, and have an open issue for APG to track that issue: https://github.com/w3c/aria-practices/issues/1156

As far as I can tell, the next step for this PR is review.

**Comment by zcorpan** [comment]
https://github.com/w3c/aria/issues/915 has been resolved, and I've now updated this PR.

File: aria-practices.html [hunk 1]
```diff
@@ -6279,6 +6279,119 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+    <p>The <code>aria-level</code> attribute can be used on the following roles:</p>
+    <ul>
+      <li><code>heading</code></li>
+      <li><code>listitem</code></li>
+      <li><code>treeitem</code></li>
+      <li><code>row</code></li>
+    </ul>
+    <section id="aria-level_heading_role">
+      <h3><code>heading</code> Role</h3>
+      <p>Used together, the <code>aria-level</code> attribute and <code>heading</code> role will be treated the same by assistive technologies as the native HTML header elements: <code>h1</code>, <code>h2</code>, <code>h3</code>, <code>h4</code>, <code>h5</code> and <code>h6</code>. Do not use attribute <code>aria-level</code> and <code>heading</code> when a native HTML element can be used.</p>
```
Comment (zcorpan) [inline aria-practices.html]: Check spelling and grammar. "Do not use the `aria-level` attribute and the `heading` role ..."

**Comment by charmarkk** [comment]
The layout of some of the code snippets seems a little wonky and it seems like they don't wrap. Should they/is that an issue of the preview?

Otherwise, looks good to me. 👌

**Comment by zcorpan** [comment]
That's a known bug of the diff preview.

File: aria-practices.html [hunk 2]
```diff
@@ -6338,6 +6338,119 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
 
   </section>
 
+  <section id="aria-level">
+    <h2>Describing Hierarchical Structure with <code>aria-level</code></h2>
+    <p>When elements have a hierarchical relationship, such as headers of sections or data in tree structures, <code>aria-level</code> is used to communicate the hierarchy between elements to assistive technologies. The value of <code>aria-level</code> is numeric, with <code>1</code> indicating the top level of the structure. The number increases for each level of nesting.</p>
```
Comment (carmacleod) [inline aria-practices.html]: Please change:
> headers of sections

to:
> section headings

(Note head*ing* and not head*er*)

File: aria-practices.html [hunk 3]
```diff
@@ -6338,6 +6338,119 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+      <li><code>row</code></li>
+    </ul>
+    <section id="aria-level_heading_role">
+      <h3><code>heading</code> Role</h3>
+      <p>Used together, the <code>aria-level</code> attribute and <code>heading</code> role will be treated the same by assistive technologies as the native HTML heading elements: <code>h1</code>, <code>h2</code>, <code>h3</code>, <code>h4</code>, <code>h5</code> and <code>h6</code>. Do not use the <code>aria-level</code> attribute and the <code>heading</code> role when a native HTML element can be used.</p>
+      <p>This example uses the <code>heading</code> role and <code>aria-level</code> attribute to communicate levels of headings for a graph created with an SVG. The headings "Deciduous Trees" and "Evergreen Trees" are both subheadings to "Total Trees".</p>
+      <pre><code>&lt;svg width="1000" height="1000"&gt;
+  &lt;text x="10" y="10" role="header" aria-level="1"&gt;Total Trees:&lt;/text&gt;
+  &lt;text x="10" y="100" role="header" aria-level="2"&gt;Deciduous Trees:&lt;/text&gt;
+  &lt;text x="10" y="200" role="header" aria-level="2"&gt;Evergreen Trees:&lt;/text&gt;
```
Comment (carmacleod) [inline aria-practices.html]: Three instances of `role="header"` in this example need to be changed to `role="heading"`.

File: aria-practices.html [hunk 4]
```diff
@@ -6338,6 +6338,119 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+  ...
+&lt;/svg&gt;
+</code></pre>
+
+      <section id="aria-level_heading_role_remediation">
+        <h4>Remediation Uses Cases</h4>
+        <p>In rare scenarios, legacy code cannot be converted to using HTML header elements. In these cases, the <code>aria-level</code> attribute and the <code>heading</code> role can be added to the web site's heading elements to communicate to assistive technologies that those elements should be treated like a native HTML section heading element.</p>
```
Comment (carmacleod) [inline aria-practices.html]: header -> heading

File: aria-practices.html [hunk 5]
```diff
@@ -6338,6 +6338,119 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+        <p>For example, "Definition of a Room" is a subsection of "Housing Specification":</p>
+        <pre><code>&lt;div role="heading" aria-level="1" class="header-big"&gt;Housing Specification&lt;/div&gt;
+&lt;div role="section" aria-labelledby="room-definition"&gt;
```
Comment (carmacleod) [inline aria-practices.html]: `section` is an abstract role, so the div would need to have `role="region"`...
HOWEVER, I question whether we really want this example (and the one below) to create *landmark* regions?
i.e. I think the `aria-labelledby` and the `role` should be removed from the div in the example directly above, and the `aria-labelledby` should be removed from the `section` element in the example directly below, and the `id="room-definition"` on the headings would no longer be needed.

@mcking65 What do you think?

File: aria-practices.html [hunk 6]
```diff
@@ -6338,6 +6338,119 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+  &lt;div role="heading" aria-level="2" id="room-definition" class="header-small"&gt;Definition of a Room&lt;/div&gt;
```
Comment (carmacleod) [inline aria-practices.html]: Change:
> `class="header-big"` and `class="header-small`

to:
> `class="heading-big"` and `class="heading-small`

I realize this is not technically required, but I think it could be confusing to some authors if we use "header" when talking about "headings".

**Comment by charmarkk** [comment]
>
> That's a known bug of the diff preview.

I had thought as much, thanks for clearing that up friendo!

**Comment by zcorpan** [comment]
Thank you for the review, @carmacleod! I've addressed your comments.

I think you're right that the remediation use case example shouldn't add landmark regions.

**Comment by jongund** [comment]
@mcking65 In section 8.2 for listitem role, wouldn't a better example be for lists that are partially rendered list do to size/length of the list and or bandwidth limitations of the device. This I believe this is one of the primary reason for other similar attributes like aria-posinset and aria-setsize to be included in ARIA.

**Comment by jongund** [comment]
@mcking65 In section 8.1 Heading Role, what if someone wants to use a aria-level value greater than 6? For example, in some type of document with a high level of nesting. Is that going to be allowed? If not we should state the allowed range i 1 to 6, if allowed there should be some type of guidance on values greater than 6.

File: aria-practices.html [hunk 7]
```diff
@@ -6338,6 +6338,115 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+        <p>Equivalent (and preferred) HTML:</p>
+        <pre><code>&lt;h1&gt;Housing Specification&lt;/h1&gt;
+&lt;h2 id="room-definition"&gt;Definition of a Room&lt;/h2&gt;
+...
+</code></pre>
+      </section>
+    </section>
+
+    <section id="aria-level_listitem_role">
+      <h3><code>listitem</code> role</h3>
```
Comment (carmacleod) [inline aria-practices.html]: Role should be capitalized, for consistency with other headings.

File: aria-practices.html [hunk 8]
```diff
@@ -6338,6 +6338,115 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+&lt;ul&gt;
+  &lt;li aria-level="2"&gt;Wash Dishes&lt;/li&gt;
+&lt;/ul&gt;
+&lt;/blockquote&gt;
+&lt;p&gt;I can do this item, but I can't do the other clean up items.&lt;/p&gt;
+</code></pre>
+    </section>
+
+    <section id="aria-level_treeitem_role">
+      <h3><code>treeitem</code> role</h3>
```
Comment (carmacleod) [inline aria-practices.html]: Role should be capitalized, for consistency with other APG headings.

**Comment by JAWS-test** [comment]
WCAG ((ARIA12)[https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA12.html]) assumes that heading arial-level can be greater than 6. See: https://github.com/w3c/aria/issues/1050

File: aria-practices.html [hunk 9]
```diff
@@ -6338,6 +6338,115 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+    <p>Do not use <code>aria-level</code> when the DOM structure already accurately represents the elements' hierarchical relationships. For example, a nested <code>ul</code> list structure does not need explicit <code>aria-level</code>.</p>
+    <p>The <code>aria-level</code> attribute can be used on the following roles:</p>
+    <ul>
+      <li><code>heading</code></li>
+      <li><code>listitem</code></li>
+      <li><code>treeitem</code></li>
+      <li><code>row</code></li>
+    </ul>
+    <section id="aria-level_heading_role">
+      <h3><code>heading</code> Role</h3>
```
Comment (zcorpan) [inline aria-practices.html]: Make headings more verbose/redundant so they can stand alone better. Like "Using aria-level for the heading Role".

File: aria-practices.html [hunk 10]
```diff
@@ -6338,6 +6338,115 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+&lt;h2 id="room-definition"&gt;Definition of a Room&lt;/h2&gt;
+...
+</code></pre>
+      </section>
+    </section>
+
+    <section id="aria-level_listitem_role">
+      <h3><code>listitem</code> role</h3>
+      <p>The <code>aria-level</code> attribute can be used on elements with <code>listitem</code> role to represent list subitems when the structure of the DOM tree does not imply the intended level of nesting. This is a rare scenario, however. For example, consider that you want to <em>quote</em> a single item from a large nested to-do list. Instead of using nested <code>ul</code> elements until the right level of nesting is reached, the <code>aria-level</code> attribute on the list item can be used, as follows:</p>
+      <pre><code>&lt;blockquote&gt;
```
Comment (zcorpan) [inline aria-practices.html]: This example was confusing. Try a different example, like a ToC where the top-level chapters are on one page and the sections for each chapter are on separate pages.

File: aria-practices.html [hunk 11]
```diff
@@ -6338,6 +6338,115 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+&lt;/blockquote&gt;
+&lt;p&gt;I can do this item, but I can't do the other clean up items.&lt;/p&gt;
+</code></pre>
+    </section>
+
+    <section id="aria-level_treeitem_role">
+      <h3><code>treeitem</code> role</h3>
+      <p>In some implementations of the <a href="#TreeView">Tree View Design Pattern</a>, authors might need to explicitly declare the <code>aria-level</code> attribute.</p>
+      <p>The computed <code>aria-level</code> of a <code>treeitem</code> within a tree is based on the number of <code>group</code> role elements in the ancestor chain between the <code>treeitem</code> and the <code>tree</code> role, where the top-level <code>treeitem</code>s are <code>aria-level</code> 1. In the following example of a discussion board with nested replies, the top-level post "What color should we paint the bike shed?" has an implicit <code>aria-level='1'</code>. The first reply has an implicit <code>aria-level='2'</code>, the response to that first reply has an implicit <code>aria-level='3'</code>.</p>
+      <pre><code>&lt;ul role='tree'&gt;
```
Comment (zcorpan) [inline aria-practices.html]: This is apparently not a proper use of `tree`.

Can anyone suggest a better example?

**Comment by zcorpan** [comment]
I have addressed the comments. 2 examples are new. Please review. 🙂

@JAWS-test ARIA doesn't put any limit to `aria-level` currently. Is the limitation in JAWS only for `heading`? Should aria-practices mention it?

**Comment by JAWS-test** [comment]
> ARIA doesn't put any limit to aria-level currently. Is the limitation in JAWS only for heading? 

Yes, only for heading.

> Should aria-practices mention it?

I don't think so, because other errors of AT are not named either. I think it would be good if the ARIA specification explicitly states that aria-level can also be greater than 6 for heading.

**Comment by zcorpan** [comment]
OK. I see that you have filed https://github.com/w3c/aria/issues/1050 already.

**Comment by zcorpan** [comment]
@carmacleod can you review this? Thanks!

File: aria-practices.html [hunk 12]
```diff
@@ -6338,6 +6338,127 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+  &lt;/li>
+  &lt;li>This is level 1&lt;/li>
+&lt;/ul>
+</code></pre>
+    <p>The <code>aria-level</code> attribute can be used on the following roles:</p>
+    <ul>
+      <li><code>heading</code></li>
+      <li><code>listitem</code></li>
+      <li><code>treeitem</code></li>
+      <li><code>row</code></li>
```
Comment (carmacleod) [inline aria-practices.html]: Maybe mention here that aria-level can only be used on row if it's in a treegrid:

```suggestion
      <li><code>row</code> (if in a <code>treegrid</code>)</li>
```

File: aria-practices.html [hunk 13]
```diff
@@ -6338,6 +6338,127 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+&lt;h2 id="intended-audience">Intended audience&lt;/h2>
+...
+</code></pre>
+
+      <p class="note">The HTML <code>li</code> element uses the <code>listitem</code> role by default.</p>
+    </section>
+
+    <section id="aria-level_treeitem_role">
+      <h3>Using <code>aria-level</code> for the <code>treeitem</code> Role</h3>
+      <p>In some implementations of the <a href="#TreeView">Tree View Design Pattern</a>, authors might need to explicitly declare the <code>aria-level</code> attribute.</p>
```
Comment (carmacleod) [inline aria-practices.html]: It feels to me like there needs to be one more sentence (in its own paragraph) before this one, that describes "normal" use of aria-level for the treeitem role. It feels like we've jumped into the exceptional case, and didn't even mention the typical case.

Comment (carmacleod) [reply aria-practices.html]: @zcorpan 
Oh. I guess aria-level is normally computed by the browser, based on DOM structure.

It would still be nice to have a short intro sentence like, "Usually, aria-level is computed by the browser based on DOM structure.", then follow with "In some implementations...".

**Comment by zcorpan** [comment]
The HTML diff service gave a 500 error when the PR preview ran. Now the service seems to be OK again, so I'll close and reopen to rerun PR preview.

File: aria-practices.html (hunk 13 above)
Comment (zcorpan) [reply aria-practices.html]: @mcking65 clarified this in a716ad878458eadeaf28e6a82b59995b75de1d55

**Comment by zcorpan** [comment]
@jongund you had some comments about the examples in https://github.com/w3c/aria-practices/pull/1109#issuecomment-550030262 . The examples have been changed since that telecon. Would you like to review? Thanks!

**Comment by jongund** [comment]
I would recommend changing the class names in the following example:

```<div role="heading" aria-level="1" class="heading-big">Housing Specification</div>
<div role="heading" aria-level="2" id="room-definition" class="heading-small">Definition of a Room</div>
````

to

<div role="heading" aria-level="1" class="main-section">Housing Specification</div>
<div role="heading" aria-level="2" id="room-definition" class="sub-section">Definition of a Room</div>

**Comment by jongund** [comment]
@zcorpan 

I think the section 8.2 should also talk more specially when only part of a longer list is rendered due to size of the list, screen real estate or bandwith issues, these examples would also include aria-posinset and aria-setsize.  This is an example is a list of potential search categories:

```
<h1>Subject Search Terms</h1>
<ul>
<li aria-level="1" aria-posinset="3" aria-setsize="21">Life Sciences</li>
<li aria-level="2" aria-posinset="1" aria-setsize=""10">Andrology</li>
<li aria-level="2" aria-posinset="2" aria-setsize=""10">Animal Physiology</li>
<li aria-level="2" aria-posinset="3" aria-setsize=""10">Biochemistry</li>
<li aria-level="2" aria-posinset="4" aria-setsize=""10">Bioinformatices</li>
</ul>
```

**Comment by zcorpan** [comment]
@jongund 

> I would recommend changing the class names in the following example:

This is under "Remediation" - the assumption is that the original markup is bad, but can't be changed other than adding ARIA. Maybe it could explain this better, by using another code snippet for what the original markup is (and that it's bad), and saying what the constraints are more clearly.

**Comment by scottaohara** [comment]
Regarding what screen readers do with levels beyond 6, [here's a test file](https://scottaohara.github.io/tests/html-headings/6plus.html) I put together back in October.

tldr; seems the levels are correctly exposed at the platform level, but can be erratic when announced by screen readers.  e.g. any heading beyond 6 announced as an `h2`

**Comment by zcorpan** [comment]
I have added a note about heading levels > 6.

I have also tried to address @jongund's comments.

**Comment by zcorpan** [comment]
@jongund, @carmacleod, can you review the latest 2 commits?

File: aria-practices.html [hunk 14]
```diff
@@ -6411,6 +6411,188 @@ <h3>Indicating sort order with <code>aria-sort</code></h3>
...
+
+      <p class="note">The implied (or default) role of an HTML <code>li</code> element is <code>listitem</code>.</p>
+
+      <p>
+        The <code>aria-level</code> property can also be used in conjuction with the <code>aria-posinset</code> and <code>aria-setsize</code> properties.
+        For example, when showing a subset of a long list, but it is still useful to communicate the original position and number of items in the list.
+      </p>
+
+      <div class="example">
+        <p>This is an example is a list of potential search categories:</p>
```
Comment (carmacleod) [inline aria-practices.html]:
```suggestion
        <p>This is an example of a list of potential search categories:</p>
```

**Comment by carmacleod** [comment]
Looks good. The note about > 6 is useful - thanks.

I see that class="example" was used, which gives the example a light orange background and a vertical orange bar in the left margin. I see this style in older versions of the APG (and only on `<pre>` elements), but it's not used in the current version. We should ask the group if it was intentionally removed.

**Comment by zcorpan** [comment]
Thanks! @mcking65 ok to merge?

File: aria-practices.html [hunk 15]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
       <pre><code>&lt;label&gt;Price per paperclip: $&lt;input type="number" min="0.01" value="0.5" max="2" step="0.01"&gt;&lt;/label&gt;</code></pre>
     </section>
   </section>
+  
+  <section id="aria-level">
+    <h2>Describing Hierarchical Structure with <code>aria-level</code></h2>
+    <p>
+      When elements have a hierarchical relationship, such as headings of nested sections or data in tree structures, the value of <a href="#aria-level" class="property-reference">aria-level</a> communicates an element's level within the hierarchy to assistive technologies.
+      The value of <code>aria-level</code> is numeric, with <code>1</code> indicating the top level of the structure.
```
Comment (carmacleod) [inline aria-practices.html:6742]: Let's make it very clear that level is an integer, and not just "numeric" (because someone might think that they can use "1.2" or similar).

```suggestion
      The value of <code>aria-level</code> is an integer greater than or equal to 1, with <code>1</code> indicating the top level of the structure.
```

File: aria-practices.html [hunk 16]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
...
+        While ARIA 1.2 supports <code>aria-level</code> on <code>listitem</code> elements, the Authoring Practices Task Force has not found a scenario where applying <code>aria-level</code> to a <code>listitem</code> is needed and has
+        <a href="https://github.com/w3c/aria/issues/1369">raised an issue</a>
+        suggesting the support be deprecated.
+      </p>
+    </section>
+
+    <section id="aria-level_treeitem_role">
+      <h3>Using <code>aria-level</code> with the <code>treeitem</code> Role</h3>
+      <p>
+        As with <code>listitem</code> elements, browsers automatically compute an <code>aria-level</code> value for elements with role <code>treeitem</code>.
```
Comment (carmacleod) [inline aria-practices.html:6829]: Let's not refer to listitem levels in the treeitem section.

```suggestion
        Browsers automatically compute an <code>aria-level</code> value for elements with role <code>treeitem</code>.
```

File: aria-practices.html [hunk 17]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
...
+  &lt;li role="treeitem" class="doc" aria-level="3">report-1B.docx&lt;/li>
+  &lt;li role="treeitem" class="doc" aria-level="3">report-1C.docx&lt;/li>
+&lt;/ul&gt;
+</code></pre>
+       <p class="note">See the <a href="#TreeView">Tree View Design Pattern</a> for how this example handles focus and keyboard navigation.</p>
+    </section>
+
+    <section id="aria-level_row_role">
+      <h3>Using <code>aria-level</code> with the <code>row</code> Role</h3>
+      <p>The <code>aria-level</code> attribute can be used on an element with <code>row</code> role to describe nesting of rows in a <code>treegrid</code>, as is explained in the <a href="#treegrid">treegrid pattern</a>. In this example, the DOM tree does not represent the hierarchical relationship between rows. Each email is in a <code>tr</code> element, which are siblings in the DOM tree, therefore <code>aria-level</code> is necessary to communicate that the emails form a tree structure.</p>
```
Comment (carmacleod) [inline aria-practices.html:6851]: Don't need "is":
```suggestion
      <p>The <code>aria-level</code> attribute can be used on an element with <code>row</code> role to describe nesting of rows in a <code>treegrid</code>, as explained in the <a href="#treegrid">treegrid pattern</a>. In this example, the DOM tree does not represent the hierarchical relationship between rows. Each email is in a <code>tr</code> element, which are siblings in the DOM tree, therefore <code>aria-level</code> is necessary to communicate that the emails form a tree structure.</p>
```

File: aria-practices.html [hunk 18]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
...
+  &lt;td&gt;Laura&lt;/td&gt;
+  &lt;td&gt;We should definitely cover the bike shed in rainbows.&lt;/td&gt;
+  &lt;/tr&gt;
+  &lt;tr aria-level='2'&gt;
+  &lt;td&gt;Fred&lt;/td&gt;
+  &lt;td&gt;Re: We should definitely cover the bike shed in rainbows.&lt;/td&gt;
+  &lt;/tr&gt;
+&lt;/table&gt;
+</code></pre>
+      <p>Do not use <code>aria-level</code> on rows that are in a <code>grid</code> or a <code>table</code> because only rows in a <code>treegrid</code> table are expected to have a hierarchical relationship.</p>
```
Comment (carmacleod) [inline aria-practices.html:6867]: Do we really need to say "treegrid table"? Maybe delete "table" and just say "treegrid" here?
```suggestion
      <p>Do not use <code>aria-level</code> on rows that are in a <code>grid</code> or a <code>table</code> because only rows in a <code>treegrid</code> are expected to have a hierarchical relationship.</p>
```

**Review by carmacleod — APPROVED** [review]
+1 as-is, but I added 4 small editorial suggestions to consider.

**Review by mcking65 — APPROVED** [review]

**Comment by jongund** [comment]
@mcking65 

Suggested edits to section 9.3:
Change:
"However, in some implementations of the Tree View Design Pattern, it is necessary to explicitly declare the aria-level."

To:
 However, in some implementations of the Tree View Design Pattern when the DOM structure does not represent the actual level and the level is important for understanding the content or actions of the treeitem, it is necessary to explicitly declare the aria-level.  

Suggested edits to section 9.4:
Add the following sentences:
The rows with aria-level=2 represent responses to the last previous row with aria-level=1 and visually this relationship is typically displayed the rows with aria-level=2 indented.  aria-level in this example is providing the information of the response relationship that is represented visually by indenting.  

Otherwise looks great!

**Review by jongund — CHANGES_REQUESTED** [review]
@mcking65 

Suggested edits to section 9.3:
Change:
"However, in some implementations of the Tree View Design Pattern, it is necessary to explicitly declare the aria-level."

To:
 However, in some implementations of the Tree View Design Pattern when the DOM structure does not represent the actual level and the level is important for understanding the content or actions of the treeitem, it is necessary to explicitly declare the aria-level.  

Suggested edits to section 9.4:
Add the following sentences:
The rows with aria-level=2 represent responses to the last previous row with aria-level=1 and visually this relationship is typically displayed the rows with aria-level=2 indented.  aria-level in this example is providing the information of the response relationship that is represented visually by indenting.  

Otherwise looks great!

**Review by a11ydoer — CHANGES_REQUESTED** [review]
@mcking65 

In Section 9.3, the code example, 

`<h3 id="tree _label">Viewing: report-1</h3>`
 should be **h2**, not h3 because 
1) The text in the description said, " the "report-1" folder is level 2." and 
2) next heading level in li element is  aria-level ="3"

**Comment by a11ydoer** [comment]
> Suggested edits to section 9.4:
> Add the following sentences:
> The rows with aria-level=2 represent responses to the last previous row with aria-level=1 and visually this relationship is typically displayed the rows with aria-level=2 indented. aria-level in this example is providing the information of the response relationship that is represented visually by indenting.
> 
Thanks, @jongund 
Now I understand why the response is aria-level="2"

**Review by a11ydoer — CHANGES_REQUESTED** [review]
9.4 Using aria-level with the row Role

In the code example, aria-level value should have double quotation, not single quotation.

File: aria-practices.html [hunk 19]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
...
+      </p>
+    </section>
+
+    <section id="aria-level_treeitem_role">
+      <h3>Using <code>aria-level</code> with the <code>treeitem</code> Role</h3>
+      <p>
+        Browsers automatically compute an <code>aria-level</code> value for elements with role <code>treeitem</code>.
+        The computed level for a <code>treeitem</code> element is based on the number of <code>group</code> elements in the ancestor chain between that <code>treeitem</code> and the element with role <code>tree</code>.
+        The top-level <code>treeitem</code> elements, which are immediate descendants of the <code>tree</code> with no intervening <code>group</code>, have <code>aria-level</code> of <code>1</code>.
+        However, in some implementations of the <a href="#TreeView">Tree View Design Pattern</a>, it is necessary to explicitly declare the <code>aria-level</code>.
```
Comment (zcorpan) [inline aria-practices.html:6832]:
```suggestion
        However, in some implementations of the <a href="#TreeView">Tree View Design Pattern</a>
        when the DOM structure does not represent the actual level and the level is important for understanding the content or actions of the treeitem,
        it is necessary to explicitly declare the <code>aria-level</code>.
```

File: aria-practices.html [hunk 20]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
...
+  &lt;li role="treeitem" class="doc" aria-level="3">report-1B.docx&lt;/li>
+  &lt;li role="treeitem" class="doc" aria-level="3">report-1C.docx&lt;/li>
+&lt;/ul&gt;
+</code></pre>
+       <p class="note">See the <a href="#TreeView">Tree View Design Pattern</a> for how this example handles focus and keyboard navigation.</p>
+    </section>
+
+    <section id="aria-level_row_role">
+      <h3>Using <code>aria-level</code> with the <code>row</code> Role</h3>
+      <p>The <code>aria-level</code> attribute can be used on an element with <code>row</code> role to describe nesting of rows in a <code>treegrid</code>, as explained in the <a href="#treegrid">treegrid pattern</a>. In this example, the DOM tree does not represent the hierarchical relationship between rows. Each email is in a <code>tr</code> element, which are siblings in the DOM tree, therefore <code>aria-level</code> is necessary to communicate that the emails form a tree structure.</p>
```
Comment (zcorpan) [inline aria-practices.html:6851]:
```suggestion
      <p>The <code>aria-level</code> attribute can be used on an element with <code>row</code> role to describe nesting of rows in a <code>treegrid</code>, as explained in the <a href="#treegrid">treegrid pattern</a>.
      In this example, the DOM tree does not represent the hierarchical relationship between rows.
      Each email is in a <code>tr</code> element, which are siblings in the DOM tree, therefore <code>aria-level</code> is necessary to communicate that the emails form a tree structure.
      The rows with <code>aria-level="2"</code> represent responses to the last previous row with <code>aria-level="1"</code> and visually this relationship is typically displayed the rows with <code>aria-level="2"</code> indented. <code>aria-level</code> in this example is providing the information of the response relationship that is represented visually by indenting.</p>
```

File: aria-practices.html [hunk 21]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
...
+        The computed level for a <code>treeitem</code> element is based on the number of <code>group</code> elements in the ancestor chain between that <code>treeitem</code> and the element with role <code>tree</code>.
+        The top-level <code>treeitem</code> elements, which are immediate descendants of the <code>tree</code> with no intervening <code>group</code>, have <code>aria-level</code> of <code>1</code>.
+        However, in some implementations of the <a href="#TreeView">Tree View Design Pattern</a>, it is necessary to explicitly declare the <code>aria-level</code>.
+      </p>
+      <p>
+        The following example shows a File Directory where a nested folder "report-1" is being viewed.
+        The parent folder "Reports" is at the top-level in the directory structure, and the "report-1" folder is level 2.
+        The files within the "report-1" folder are thus level 3, even though only these files are currently being viewed.
+      </p>
+      <pre><code>&lt;h3 id="tree_label">Viewing: report-1&lt;/h3>
```
Comment (zcorpan) [inline aria-practices.html:6839]:
```suggestion
      <pre><code>&lt;h2 id="tree_label">Viewing: report-1&lt;/h2>
```

File: aria-practices.html [hunk 22]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
...
+&lt;ul role='tree' aria-labelledby='tree_label'&gt;
```
Comment (zcorpan) [inline aria-practices.html:6840]:
```suggestion
&lt;ul role="tree" aria-labelledby="tree_label"&gt;
```

File: aria-practices.html [hunk 23]
```diff
@@ -6737,6 +6734,139 @@ <h3>Range properties with spin buttons</h3>
...
+      <pre><code>&lt;table role='treegrid'&gt;
+  &lt;tr&gt;
+  &lt;th&gt;From:&lt;/th&gt;
+  &lt;th&gt;Subject:&lt;/th&gt;
+  &lt;/tr&gt;
+  &lt;tr aria-level='1'&gt;
+  &lt;td&gt;Laura&lt;/td&gt;
+  &lt;td&gt;We should definitely cover the bike shed in rainbows.&lt;/td&gt;
+  &lt;/tr&gt;
+  &lt;tr aria-level='2'&gt;
```
Comment (zcorpan) [inline aria-practices.html:6861]:
```suggestion
      <pre><code>&lt;table role="treegrid"&gt;
  &lt;tr&gt;
  &lt;th&gt;From:&lt;/th&gt;
  &lt;th&gt;Subject:&lt;/th&gt;
  &lt;/tr&gt;
  &lt;tr aria-level="1"&gt;
  &lt;td&gt;Laura&lt;/td&gt;
  &lt;td&gt;We should definitely cover the bike shed in rainbows.&lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr aria-level="2"&gt;
```

**Comment by zcorpan** [comment]
@mcking65 @jongund @a11ydoer thanks, addressed.
